ghcide/src/Development/IDE/Types/Exports.hs

```diff
@@ -90,26 +97,44 @@ mkIdentInfos mod (AvailTC _ nn flds)
         | n <- nn ++ map flSelector flds
       ]
 
+buildModuleExportMap:: [ModIface] -> Map.HashMap Text (HashSet IdentInfo)
+buildModuleExportMap modIfaces = do
+  let exports = map extractModuleExports modIfaces
+  Map.fromListWith (<>) exports
+
+extractModuleExports :: ModIface -> (Text, HashSet IdentInfo)
+extractModuleExports modIFace = do
+  let modName = pack $ moduleNameString $ moduleName $ mi_module modIFace
+  let functionSet = Set.fromList $ concatMap (mkIdentInfos modName) $ mi_exports modIFace
+  (modName, functionSet)
+
 createExportsMap :: [ModIface] -> ExportsMap
-createExportsMap = ExportsMap . Map.fromListWith (<>) . concatMap doOne
+createExportsMap modIface = do
+  let exportsMap = (Map.fromListWith (<>) . concatMap doOne) modIface
+      moduleExportMap = buildModuleExportMap modIface
+  ExportsMap exportsMap moduleExportMap
   where
-    doOne mi = concatMap (fmap (second Set.fromList) . unpackAvail mn) (mi_exports mi)
-      where
-        mn = moduleName $ mi_module mi
+    doOne mi = do
+      let getModDetails = unpackAvail $ moduleName $ mi_module mi
+      concatMap (fmap (second Set.fromList) . getModDetails) (mi_exports mi)
```

Comment:
Can you find a way to rewrite this code so that the `IdentInfo` values in both hashmaps? The duplication is expensive both in time and space.

This comment applies to all the `createExportsMap` functions in this module.

Reply:
I meant to say "the values are reused in both hash maps". My keyboard ate a word 😅

Reply:
One simple way to do this is to generate it from the already existing map.. did you mean a different approach? This would also make the below comment simple

Reply:
Generating from the existing map is fine. Since exports are already organized by module, it might be more effective to generate the module exports map first, and derive the other one from it.

Reply:
resolved - I now generate both from a list